oxide-vpc/src/engine/icmpv6.rs

```diff
@@ -9,41 +9,169 @@
 use opte::api::{Direction, OpteError};
+use opte::engine::ether::ETHER_TYPE_IPV6;
 use opte::engine::icmpv6::Icmpv6EchoReply;
+use opte::engine::icmpv6::MessageType;
+use opte::engine::icmpv6::NeighborAdvertisement;
+use opte::engine::icmpv6::RouterAdvertisement;
 use opte::engine::layer::Layer;
 use opte::engine::port::{PortBuilder, Pos};
+use opte::engine::rule::DataPredicate;
+use opte::engine::rule::EtherTypeMatch;
+use opte::engine::rule::IpProtoMatch;
+use opte::engine::rule::Predicate;
 use opte::engine::rule::{Action, Rule};
+use smoltcp::wire::Icmpv6Message;
 
 pub fn setup(
     pb: &mut PortBuilder,
     cfg: &VpcCfg,
-    ft_limit: core::num::NonZeroU32,
-) -> core::result::Result<(), OpteError> {
-    // The ICMPv6 layer only contains meaningful actions if the port is
-    // configured to support IPv6.
-    let ip_cfg = match cfg.ipv6_cfg() {
-        None => return Ok(()),
-        Some(cfg) => cfg,
-    };
-
-    let reply = Action::Hairpin(Arc::new(Icmpv6EchoReply {
-        // Map an Echo Request from guest (src) -> gateway (dst) to an Echo
-        // Reply from gateway (dst) -> guest (src).
-        src_mac: cfg.private_mac.into(),
-        src_ip: ip_cfg.private_ip,
-        dst_mac: cfg.gateway_mac.into(),
-        dst_ip: ip_cfg.gateway_ip,
-    }));
-    let mut icmp = Layer::new("icmpv6", pb.name(), vec![reply], ft_limit);
-
-    let rule = Rule::new(1, icmp.action(0).unwrap().clone());
+    ft_limit: NonZeroU32,
+) -> Result<(), OpteError> {
+    match cfg.ipv6_cfg() {
+        None => drop_all_icmpv6(pb, ft_limit),
+        Some(ip_cfg) => add_icmpv6_rules(pb, cfg, ip_cfg, ft_limit),
+    }
+}
+
+// Explicitly drop any ICMPv6 traffic if the guest is not configured with an
+// IPv6 address.
+fn drop_all_icmpv6(
+    pb: &mut PortBuilder,
+    ft_limit: NonZeroU32,
+) -> Result<(), OpteError> {
+    let mut rule = Rule::new(1, Action::Deny);
+    rule.add_predicate(Predicate::InnerEtherType(vec![EtherTypeMatch::Exact(
+        ETHER_TYPE_IPV6,
+    )]));
+    rule.add_predicate(Predicate::InnerIpProto(vec![IpProtoMatch::Exact(
+        Protocol::ICMPv6,
+    )]));
+    let rule = rule.finalize();
+
+    let mut icmp = Layer::new("icmpv6", pb.name(), vec![], ft_limit);
+    icmp.add_rule(Direction::In, rule.clone());
+    icmp.add_rule(Direction::Out, rule.clone());
+    pb.add_layer(icmp, Pos::Before("firewall"))
+}
+
+// Add support for ICMPv6:
+//
+// - Respond to echo requests from the guest to the gateway. The source address
+// may be either any link-local address in guest (since we can't know how they
+// generate that) or its assigned VPC-private address. The destination address
+// must be the link-local address we derive for OPTE, from the EUI-64 transform
+// on its MAC address.
+//
+// - Respond to NDP Router Solicitations from the guest to the gateway.
+//
+// - Respond to NDP Neighbor Solicitations from the guest to the gateway, with a
+// Target Address of both the link-local and VPC-private IPv6 address.
```

Comment:
This comment has me confused. We are responding to a unicast solicitation for the Target Address (which I think is always going to be the router's link-local address if I'm reading things correctly) and we respond with the corresponding target link-layer (MAC) address.

Reply:
Yeah, that's just wrong. I had something similar when I supported pings from the VPC-private IPv6 address of the client. I'll clean it up now.